core/store.go

```diff
@@ -277,6 +279,119 @@ func (store *Store) Get(k string) *Obj {
 	return store.getHelper(k, true)
 }
 
+// CloneKeypool creates and returns a deep copy of the store's keypool.
+// The keypool is a map that associates string keys with pointers to objects.
+// The function locks the keypool for reading to ensure thread-safe access while copying.
+//
+// Returns:
+// - clone: A new map containing the same keys and pointers as the original keypool.
+func (store *Store) CloneKeypool() map[string]*string {
+	// Lock the keypool for reading to ensure thread safety while copying.
+	store.keypoolMutex.RLock()
+	defer store.keypoolMutex.RUnlock()
+
+	clone := make(map[string]*string, len(store.keypool))
+
+	for k, v := range store.keypool {
+		clone[k] = v
+	}
+
+	return clone
+}
+
+// scanKeys performs a scan operation on the keypool, retrieving keys that match specified criteria.
+// The scan operation is controlled by a cursor and can be limited by a count, pattern, and key type.
+// The function takes a snapshot of the keypool to ensure that the scan is performed on a consistent view of the data.
+//
+// Arguments:
+// - cursor: An integer representing the current scan position within the keypool.
+// - count: The maximum number of keys to return in this scan operation.
+// - pattern: A string representing a glob pattern to filter keys (optional).
+// - keyType: A string representing the type of keys to include in the results (optional).
+//
+// Returns:
+// - newCursor: The updated cursor position after the scan, or 0 if the end is reached.
+// - keys: A slice of strings containing the keys that match the scan criteria.
+func (store *Store) scanKeys(cursor, count int, pattern, keyType string) (newCursor int, keys []string) {
+	snapshot := store.CloneKeypool()
```

Comment:
So we are iterating twice over the key-set, once while cloning the keypool and again iterating through the cloned pool while scanning for keys.
If the number ok keys are very high, it may result in high latency...

Reply:
So for this I can use single iteration in scanning itself like don't clone it means remove the clonekeypool function this will reduce the latency

```go
store.keypoolMutex.RLock()
defer store.keypoolMutex.RUnlock()

keyList := make([]string, 0, len(store.keypool))
for k := range store.keypool {
    keyList = append(keyList, k)
}
```